[SwiftParser] Improve ergonomics for initializers of `RawXXXSyntax`

AppAppWorks opened the pull request:
This PR aims to improve developer ergonomics when using the designated initializers of raw syntax nodes. Simply put, it's through achieving feature parity with the non-raw side.

### Prefer generic types over type erasure types

Currently the designated initializer of `RawXXXSyntax` always uses type-erased types instead of generic types as parameter types, leading to much clunkiness, e.g.

```swift
public struct RawPatternExprSyntax: RawExprSyntaxNodeProtocol {
  public init(
    _ unexpectedBeforePattern: RawUnexpectedNodesSyntax? = nil, 
    pattern: RawPatternSyntax, // type erasure
    _ unexpectedAfterPattern: RawUnexpectedNodesSyntax? = nil, 
    arena: __shared SyntaxArena
  )
}

let pattern = RawIdentifierPatternSyntax(
  identifier: identifier,
  arena: self.arena
)
return RawExprSyntax(RawPatternExprSyntax(pattern: RawPatternSyntax(pattern), arena: self.arena))
```

The initializer body, however, only invokes `pattern.raw`, therefore we could just replace `pattern: RawPatternSyntax` with `pattern: some RawPatternSyntaxNodeProtocol` transparently,

```swift
public struct RawPatternExprSyntax: RawExprSyntaxNodeProtocol {
  public init(
    _ unexpectedBeforePattern: RawUnexpectedNodesSyntax? = nil, 
    pattern: some RawPatternSyntaxNodeProtocol, // generic type
    _ unexpectedAfterPattern: RawUnexpectedNodesSyntax? = nil, 
    arena: __shared SyntaxArena
  )
}

let pattern = RawIdentifierPatternSyntax(
  identifier: identifier,
  arena: self.arena
)
// clunky type-erasure removed
return RawExprSyntax(RawPatternExprSyntax(pattern: pattern, arena: self.arena))
```

Actually it's what `PatternExprSyntax.init` has been doing,
```swift
public init(
  leadingTrivia: Trivia? = nil,
  _ unexpectedBeforePattern: UnexpectedNodesSyntax? = nil,
  pattern: some PatternSyntaxProtocol,
  _ unexpectedAfterPattern: UnexpectedNodesSyntax? = nil,
  trailingTrivia: Trivia? = nil
)
```

### Generate convenience generic initializers for choices enum

e.g.
```swift
public struct RawCodeBlockItemSyntax: RawSyntaxNodeProtocol {
  public enum Item: RawSyntaxNodeProtocol {
    case `decl`(RawDeclSyntax)
    case `stmt`(RawStmtSyntax)
    case `expr`(RawExprSyntax)

    public init(_ other: some RawDeclSyntaxNodeProtocol) {
      self = .decl(RawDeclSyntax(other))
    }
    
    public init(_ other: some RawStmtSyntaxNodeProtocol) {
      self = .stmt(RawStmtSyntax(other))
    }
    
    public init(_ other: some RawExprSyntaxNodeProtocol) {
      self = .expr(RawExprSyntax(other))
    }
```

So call sites might get rid of type erasure as follows,
```swift
// requires type erasure
.expr(RawExprSyntax(RawMissingExprSyntax(arena: self.arena)))

// supports generic type 
.init(RawMissingExprSyntax(arena: self.arena))
```

### Remove type erasure in call sites

This PR has removed unnecessary uses of type-erased types in many call sites thanks to the new initializers.

ahoppen commented on CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/RawSyntaxNodesFile.swift:
```diff
@@ -270,13 +290,23 @@ func rawSyntaxNodesFile(nodesStartingWith: [Character]) -> SourceFileSyntax {
+    if !kind.isNodeChoicesEmpty {
+      paramType = "\(syntaxChoicesType)"
+    } else if hasBaseType && !isOptional {
+      paramType = "some \(syntaxNodeKind.rawProtocolType)"
```
Should we also implement this for optional nodes with a base type using `(some XXXX)?`?

AppAppWorks replied:
If we replace `Raw<Kind>Syntax?` with `(some Raw<Kind>SyntaxNodeProtocol)?`, call sites will no longer be able to just pass `nil` to this parameter because we'll need to specialize the generic type.

A workaround for this is what `XXX<Kind>Syntax` are doing,  e.g.
```swift
type: (some TypeSyntaxProtocol)? = TypeSyntax?.none,
```
I didn't know whether it'd be desirable to change all parameters like this into optional parameters for `RawXXX<Kind>Syntax`, so I've kept them as-is.

ahoppen replied:
I see. I think requiring the explicit `nil` for raw syntax makes sense because it prompts you to think whether you should parse that part of the node. Could you add this explanation as a comment?

ahoppen commented on CodeGeneration/Sources/generate-swift-syntax/templates/swiftsyntax/RawSyntaxNodesFile.swift:
```diff
@@ -270,13 +290,23 @@ func rawSyntaxNodesFile(nodesStartingWith: [Character]) -> SourceFileSyntax {
 
-    return buildableType.optionalWrapped(type: paramType)
+    if isOptional {
+      if paramType.is(SomeOrAnyTypeSyntax.self) {
```
Why is this special case needed?

ahoppen commented on Sources/SwiftSyntax/generated/raw/RawSyntaxNodesAB.swift:
```diff
@@ -42,6 +42,14 @@ public struct RawAccessorBlockSyntax: RawSyntaxNodeProtocol {
       return nil
     }
+    
+    public init(_ other: RawAccessorDeclListSyntax) {
+      self = .accessors(other)
+    }
```
Why are these initializers needed? Couldn’t we just continue using the case names?

AppAppWorks replied:
True, I'll change the code generation logic so only enum cases with the payload type `Raw<Kind>Syntax` will get a convenience `init`.

ahoppen reviewed (commented):
Thank you. That does simplify the construction of raw syntax nodes.

I am not personally a huge fan of using `.init` to create objects. It’s easy to write but I think it hurts readability along the way because you don’t know which object is being constructed. Could you change those to the actual type names or stick with the enum case name if we don’t actually need to go through the initializer?

AppAppWorks commented:
>Could you change those to the actual type names or stick with the enum case name if we don’t actually need to go through the initializer?

I've reverted most changes back to their original states, however for the convenience initializers I've implemented another idea, e.g.
```swift
/// convenience init for `decl`
public init(decl: some RawDeclSyntaxNodeProtocol) {
  self = .decl(RawDeclSyntax(decl))
}
``` 
And the call sites are more readable now?
```swift
item: .init(expr: RawMissingExprSyntax(arena: self.arena)),
```

ahoppen commented on Sources/SwiftSyntax/generated/raw/RawSyntaxNodesTUVWXYZ.swift:
```diff
@@ -2545,6 +2545,11 @@ public struct RawYieldStmtSyntax: RawStmtSyntaxNodeProtocol {
       return nil
     }
+    
+    /// convenience init for `single`
```
I’m not sure this doc comment adds much. I’d say we can remove it.

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -1793,7 +1790,7 @@ extension Parser {
         }
 
-        parameterClause = .simpleInput(RawClosureShorthandParameterListSyntax(elements: params, arena: self.arena))
+        parameterClause = .init(RawClosureShorthandParameterListSyntax(elements: params, arena: self.arena))
```
Should this be `.init(simpleInput:)` or continue being `.simpleInput`?

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -2137,11 +2134,11 @@ extension Parser {
     if elseKeyword != nil {
       if self.at(.keyword(.if)) {
-        elseBody = .ifExpr(
+        elseBody = .init(
```
Should this continue being `ifExpr`?

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -2137,11 +2134,11 @@ extension Parser {
         )
       } else {
-        elseBody = .codeBlock(self.parseCodeBlock(introducer: ifKeyword))
+        elseBody = .init(self.parseCodeBlock(introducer: ifKeyword))
```
Same here

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -1765,7 +1762,7 @@ extension Parser {
           parser.parseClosureParameter()
         }
-        parameterClause = .parameterClause(params)
+        parameterClause = .init(params)
```
Should this continue being `.parameterClause`?

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows
